oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/tools/IndexComparator.java

```diff
@@ -0,0 +1,247 @@
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.store.SimpleFSDirectory;
+import org.apache.lucene.util.BytesRef;
+
+/**
+ * Tool to compare Lucene indexes.
+ * Indexes are considered equal if everything except the following is equal:
+ * - checksums of .si, .cfe, .cfs files
+ * - size of .si, .cfe, .cfs files (up to 1000 bytes or 0.1% difference is acceptable)
+ */
+public class IndexComparator {
+
+    public static void main(String... args) throws Exception {
+        if (args.length < 1) {
+            System.out.println("Usage: java " + IndexComparator.class.getName() +
+                    " <lucene index directory> [<second lucene index directory>]");
+            System.out.println(
+                    "If only one directory is specified, then the output is json " +
+                    "that contains the statistics of the index.");
+            System.out.println(
+                    "Instead of a directory, each argument can be a json file " +
+                    "that contains the statistics of a previous run.");
+            return;
+        }
+        String stats1 = getStats(new File(args[0]));
+        if (args.length > 1) {
+            String stats2 = getStats(new File(args[1]));
+            System.out.println(diff(stats1, stats2));
+        } else {
+            System.out.println(stats1);
+        }
+    }
+
+    private static String diff(String stats1, String stats2) {
+        JsonObject o1 = JsonObject.fromJson(stats1, true);
+        JsonObject o2 = JsonObject.fromJson(stats2, true);
+        o1.getProperties().remove("directory");
+        o2.getProperties().remove("directory");
+        JsopBuilder result = new JsopBuilder();
+        result.object();
+        diff(o1, o2, "", result);
+        result.endObject();
+        if (result.toString().equals("{}")) {
+            return "";
+        }
+        return JsopBuilder.prettyPrint(result.toString());
+    }
+
+    private static void diff(JsonObject o1, JsonObject o2, String path, JsopBuilder result) {
+        boolean ignoreChecksums = false;
+        if (path.startsWith("/files/")) {
+            if (path.endsWith("index-details.txt")) {
+                // ignore index details
+                ignoreChecksums = true;
+            } else if (path.endsWith(".si")) {
+                // ".si" files contain a timestamp
+                ignoreChecksums = true;
+            } else if (path.endsWith(".cfs")
+                    || path.endsWith(".cfe")) {
+                // ".cfs" and ".cfe files depends on
+                // the aggregation order
+                ignoreChecksums = true;
+            }
+        }
+        Set<String> both = new HashSet<>(o1.getProperties().keySet());
+        both.addAll(o2.getProperties().keySet());
+        for (String k : both) {
+            String v1 = o1.getProperties().get(k);
+            String v2 = o2.getProperties().get(k);
+            if (v1 == null || v2 == null || !v1.equals(v2)) {
```

Comment:
this can be simplified

```suggestion
            if (v1 == null || !v1.equals(v2)) {
```